### ISSUE TITLE
Mount options for `/` (rootfs) in `/etc/fstab` not being applied.

### ISSUE THREAD

**Issue by griznog** [systemd/systemd#21618]
**systemd version the issue has been seen with**
 > systemd-239-51.el8.x86_64

**Used distribution**
 > Rocky Linux 8.5

**Linux kernel version used** (`uname -a`)
 > `4.18.0-348.2.1.el8_5.x86_64`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > Mount options in `/etc/fstab` applied to / during boot.

**Unexpected behaviour you saw**
 > No change to `/` after the remount.

**Steps to reproduce the problem**
1. Build a chroot filesystem with a full OS install.
2. In chroot, symlink `/init` -> `/usr/lib/systemd/systemd` (`ln -s /usr/lib/systemd/systemd /init`)
3. Add mount options to fstab:
    ```
    rootfs                                     /             tmpfs   size=128G,defaults 0 0
    ```
4. Pack chroot into an initrd/initramfs image (`find . | cpio --quiet -o -H newc | /usr/bin/pigz -c > initrd.img.gz`)
5. PXE boot server with kernel and initrd from above step using ipxe (https://ipxe.org). 
    ```
kernel --name kernel http://server/kernel        
imgextract --name container http://server/initrd.img.gz
boot kernel initrd=container quiet crashkernel=no vga=791 root=tmpfs rootfstype=tmpfs
    ```

**Additional program output to the terminal or log subsystem illustrating the issue**

Here's some relevant output from a node successfully booted this way showing that rootfs doesn't get the options applied by a remount.

```
[I am root!@login02:~]# systemctl status systemd-remount-fs.service
● systemd-remount-fs.service - Remount Root and Kernel File Systems
   Loaded: loaded (/usr/lib/systemd/system/systemd-remount-fs.service; static; vendor preset: disabled)
   Active: active (exited) since Fri 2021-12-03 17:17:36 EST; 4min 19s ago
     Docs: man:systemd-remount-fs.service(8)
           https://www.freedesktop.org/wiki/Software/systemd/APIFileSystems
 Main PID: 2358 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 3355442)
   Memory: 0B
   CGroup: /system.slice/systemd-remount-fs.service

Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
[I am root!@login02:~]# df -h /
Filesystem      Size  Used Avail Use% Mounted on
none               0     0     0    - /
[I am root!@login02:~]# stat /
  File: /
  Size: 0         	Blocks: 0          IO Block: 4096   directory
Device: 2h/2d	Inode: 2           Links: 18
Access: (0555/dr-xr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-12-03 17:17:40.108418824 -0500
Modify: 2021-12-03 17:17:36.673999992 -0500
Change: 2021-12-03 17:17:36.673999992 -0500
 Birth: -
[I am root!@login02:~]# mount | grep rootfs
none on / type rootfs (rw)
[I am root!@login02:~]# grep rootfs /etc/fstab
rootfs                                     /             tmpfs   size=128G,defaults 0 0
```

For context the goal here is to boot a simple chroot/container image without an intermediate step or switch_root, using `systemd` directly from the initramfs image as mounted by the kernel. I'm not sure if this is even possible? Pointers to where to dig deeper are appreciated. Having root mounted this way works fine but reports `/` as having size 0 and no free space, which breaks things like installing rpms even though space is available. 

griznog

**Comment by poettering** [comment]
I figure your rootfs is ramfs, not tmpfs...

**Comment by griznog** [comment]
It may be, I'm not sure how to tell. But also, I can't figure out why it would be ramfs. 

1. I added `initrd=initrd.magic` to the kernel command line, which ipxe.org implies will force the use of initramfs in UEFI.
2. I have `root=tmpfs rootfstype=tmpfs` set on the kernel command line. 
3. The kernel has all the requisite TMPFS things configured:
    ```
# grep -i tmpfs /lib/modules/4.18.0-348.2.1.el8_5.x86_64/build/.config 
CONFIG_DEVTMPFS=y
CONFIG_DEVTMPFS_MOUNT=y
CONFIG_TMPFS=y
CONFIG_TMPFS_POSIX_ACL=y
CONFIG_TMPFS_XATTR=y
    ```

Is there something else missing that I need to do to have the kernel use a `tmpfs` for the initramfs image?

**Comment by poettering** [comment]
`stat -f /` will tell you

**Comment by griznog** [comment]
```
# stat -f /
  File: "/"
    ID: 0        Namelen: 255     Type: ramfs
Block size: 4096       Fundamental block size: 4096
Blocks: Total: 0          Free: 0          Available: 0
Inodes: Total: 0          Free: 0
```

Well, that's frustrating. I'm not sure where to go from there, as I understand the assorted pages I've read on what the kernel does at boot I've checked all the boxes for this to be a `tmpfs`.

**Comment by griznog** [comment]
Since/if I can't get the kernel to make this a `tmpfs`, is there a way with `systemd` to set up a script of my own to prep a `tmpfs` to do the switch_root to and have `systemd` do that without an intermediate `/init` hack that does a `switch_root` to `systemd`? Would it be as simple as naively putting an ExectStartPre on `initrd-switch-root.service` to set things up?

**Comment by griznog** [comment]
Turns out it was something simple, more empirical testing revealed that If I change my `/etc/fstab` entry for `/` from 

```
rootfs                                     /             tmpfs   size=128G,defaults 0 0
```

to

```
tmpfs                                      /             tmpfs   size=128G,defaults 0 0
```

Then `/` is mounted as a `tmpfs` and mount options get applied correctly. The mode and gid of / are wonky, but I'm assuming I broke that.

**Comment by griznog** [comment]
Re-opening this as it turns out using `tmpfs` in `/etc/fstab` only partially fixed the issue. While `stat -f` does report `/` as ` `tmpfs` and it has a size now, the `systemd-remount-fs.service` fails:

```
# systemctl status systemd-remount-fs.service
● systemd-remount-fs.service - Remount Root and Kernel File Systems
   Loaded: loaded (/usr/lib/systemd/system/systemd-remount-fs.service; static; vendor preset: disabled)
   Active: failed (Result: exit-code) since Tue 2021-12-07 07:56:25 EST; 5s ago
     Docs: man:systemd-remount-fs.service(8)
           https://www.freedesktop.org/wiki/Software/systemd/APIFileSystems
  Process: 50484 ExecStart=/usr/lib/systemd/systemd-remount-fs (code=exited, status=1/FAILURE)
 Main PID: 50484 (code=exited, status=1/FAILURE)

Dec 07 07:56:25 login01 systemd[1]: Starting Remount Root and Kernel File Systems...
Dec 07 07:56:25 login01 systemd-remount-fs[50484]: mount: /: mount point not mounted or bad option.
Dec 07 07:56:25 login01 systemd-remount-fs[50484]: /usr/bin/mount for / exited with exit status 32.
Dec 07 07:56:25 login01 systemd[1]: systemd-remount-fs.service: Main process exited, code=exited, status=1/FAILURE
Dec 07 07:56:25 login01 systemd[1]: systemd-remount-fs.service: Failed with result 'exit-code'.
Dec 07 07:56:25 login01 systemd[1]: Failed to start Remount Root and Kernel File Systems.
```

`/` is mounted though, either `rootfs` or `tmpfs` depending on how I ask:

```
# mount | grep rootfs
rootfs on / type rootfs (rw,size=513483704k,nr_inodes=128370926)
# stat -f /
  File: "/"
    ID: 0        Namelen: 255     Type: tmpfs
Block size: 4096       Fundamental block size: 4096
Blocks: Total: 128370926  Free: 124481098  Available: 124481098
Inodes: Total: 128370926  Free: 128023698
# stat /
  File: /
  Size: 500       	Blocks: 0          IO Block: 4096   directory
Device: 2h/2d	Inode: 2           Links: 19
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-12-06 13:22:24.115797981 -0500
Modify: 2021-12-06 13:22:20.679999999 -0500
Change: 2021-12-07 07:56:25.561396249 -0500
 Birth: -
```
An attempt to manually remount and change the size also fails:
```
# mount -o remount,size=128G /
mount: /: mount point not mounted or bad option.
```
I can resize other `tmpfs` filesystems like this with no issues.

**Comment by griznog** [comment]
Again a brute force search of permutations found the error. A `tmpfs` `rootfs` does not support `acl` option, only `seclabel` and `size=` so removing `acl` from the mount options makes `systemd-remount-fs.service` succeed. Not just succeed, but succeed in a weirdly magical way that also allows subsequent `mount -o remount,size=... /` work normally as well.

For anyone else encountering this (likely a future me when I forget what I did), the working recipe for using the initrd as the final filesystem for me (Rocky Linux 8.5) was:

ipxe script:

```
kernel --name kernel http://mypxe_server_url/kernel
# imgextract to work around a kernel issue decompressing images > 4 GB.
imgextract --name container http://mypxe_server_url/initrd.img.gz
# I think rootflags is actually ignored here?
boot kernel initrd=initrd.magic initrd=container rootfs=tmpfs rootfstype=tmpfs rootflags=size=128G,seclabel 
```

And in the container, `/etc/fstab` has:

```
tmpfs           /          tmpfs   seclabel,size=128G,defaults 0 0
```

And last, in the container there is a symlink `/init -> /usr/lib/systemd/systemd` which seems to be required, even passing `init=/sbin/init` to the kernel does not seem to work when provisioning an initrd image via ipxe.
